### Retention Problem:

In [1]:
import pandas as pd

In [2]:
# Loading User Data
users_data = pd.read_csv("Database/users.csv")
users_copy = users_data.copy()

#Loading Enrollment Data
enrollment_data = pd.read_csv("Database/enrollments.csv")
enrollment_copy = enrollment_data.copy()

#Loading Session Data
session_data = pd.read_csv("Database/sessions.csv")
session_copy = session_data.copy()

# Loading Course data
course_data = pd.read_csv("Database/courses.csv")
course_copy = course_data.copy()

In [3]:
#Knowing about Datasets

print("Users Data")
print(users_data.shape)
print(users_data.columns)
print(users_data.head())

print("-------------")

print("Session Data")
print(session_data.shape)
print(session_data.columns)
print(session_data.head())

print("---------------")

print("Enrollment Data")
print(enrollment_data.shape)
print(enrollment_data.columns)
print(enrollment_data.head())


print("---------------")

print("Courses Data")
print(course_data.shape)
print(course_data.columns)
print(course_data.head())

Users Data
(500, 5)
Index(['user_id', 'name', 'age', 'location', 'signup_date'], dtype='object')
  user_id             name   age   location signup_date
0   U1000      Jodi Taylor  19.0  Bangalore  2023-09-03
1   U1001  Angela Jones MD  32.0      Delhi  2024-09-20
2   U1002   Robert Jackson  27.0    Chennai  2024-04-24
3   U1003     Michael Luna  23.0       Pune  2024-02-15
4   U1004    Renee Hawkins  20.0     Mumbai  2025-05-16
-------------
Session Data
(2000, 3)
Index(['user_id', 'session_start', 'duration_minutes'], dtype='object')
  user_id        session_start  duration_minutes
0   U1334  2024-08-21T10:03:13              10.0
1   U1091  2025-04-10T16:54:39              45.0
2   U1017  2025-01-23T23:45:39              30.0
3   U1353  2025-04-30T16:08:46              60.0
4   U1062  2024-09-10T08:54:25               NaN
---------------
Enrollment Data
(1000, 4)
Index(['user_id', 'course_id', 'enrollment_date', 'status'], dtype='object')
  user_id course_id enrollment_date     statu

In [4]:
#Checking Null Values

print("Users Data")
print(users_data.isnull().sum())

print("\n\nSession Data")
print(session_data.isnull().sum())

print("\n\nEnrollment Data")
print(enrollment_data.isnull().sum())

print("\n\nCourses Data")
print(course_data.isnull().sum())

Users Data
user_id         0
name           26
age            29
location       52
signup_date     0
dtype: int64


Session Data
user_id               0
session_start         0
duration_minutes    242
dtype: int64


Enrollment Data
user_id             0
course_id           0
enrollment_date    47
status             47
dtype: int64


Courses Data
course_id      0
course_name    0
category       5
level          2
rating         3
dtype: int64


In [5]:
#Handling Null Values

#Users Data
print("Users Data")
users_data["age_missing"] = users_data["age"].isnull().astype("int") #So in time of prediction, model can learn users with missing values behave differently
users_data["age"].fillna(users_data["age"].median(),inplace=True) #Null Values filled with Median
print("Null Values:\n",users_data.isnull().sum())

#Session Data
print("\n\n Session Data")
print("Describe:\n",session_data.describe()) # min =5 minutes and max = 120 minutes, so No Outliers
session_data["duration_minutes"].fillna(session_data["duration_minutes"].median(),inplace=True)
print("Null Values:\n",session_data.isnull().sum())

#Enrollment Data
print("\n\nEnrollment Data:")
print(enrollment_data[enrollment_data["enrollment_date"].isnull() & enrollment_data["status"].isnull()]) # gives Empty dataframe so we can drop the rows with status are null

enrollment_data = enrollment_data[~enrollment_data["status"].isnull()]
print("Describe:\n",enrollment_data.describe())
print("Null Values:\n",enrollment_data.isnull().sum())

#Need to know how much data we loss before dropping rows
before = enrollment_data.shape[0]

enrollment_data = enrollment_data[~enrollment_data["enrollment_date"].isnull()]

after = enrollment_data.shape[0]
print(before,after)
print("% of Dropped data in enrollment dataset: ",((before-after)/after)*100)

#Course Data
print("\n\nCourse Data : \n\n",course_data[course_data['category'].isnull()])
course_data = course_data[~course_data['category'].isnull()]


Users Data
Null Values:
 user_id         0
name           26
age             0
location       52
signup_date     0
age_missing     0
dtype: int64


 Session Data
Describe:
        duration_minutes
count       1758.000000
mean          40.881684
std           37.391341
min            5.000000
25%           10.000000
50%           30.000000
75%           60.000000
max          120.000000
Null Values:
 user_id             0
session_start       0
duration_minutes    0
dtype: int64


Enrollment Data:
Empty DataFrame
Columns: [user_id, course_id, enrollment_date, status]
Index: []
Describe:
        user_id course_id enrollment_date   status
count      953       953             906      953
unique     422        20             340        3
top      U1011      C203      2025-07-17  dropped
freq         7        57               8      331
Null Values:
 user_id             0
course_id           0
enrollment_date    47
status              0
dtype: int64
953 906
% of Dropped data in enrollment da

C:\Users\Avinash EM\AppData\Local\Temp\ipykernel_22364\307713898.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  users_data["age"].fillna(users_data["age"].median(),inplace=True) #Null Values filled with Median
C:\Users\Avinash EM\AppData\Local\Temp\ipykernel_22364\307713898.py:12: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on whic

##### Justification
- Enrollment Data : Data loss is minimal (5.18%) and unlikely to bias results so we can drop it
- Course Data : Since missing values were minimal (<1%) and no pattern was observed, rows were removed to maintain data integrity without impacting overall analysis

In [6]:
#Checking Duplicates

print("User Data:",users_data.duplicated().sum())
print("Session Data:", session_data.duplicated().sum())
print("Enrollment Data:", enrollment_data.duplicated().sum())
print("Course Data: ",course_data.duplicated().sum()) 

User Data: 0
Session Data: 1
Enrollment Data: 0
Course Data:  0


In [7]:
session_data[session_data.duplicated()] 

,user_id,session_start,duration_minutes
763,U1470,unknown,30.0


##### This duplicate appears to be data login error since the session_start should always auto-record. The value 'unknown' makes this data unusable for time based engagement metrix like sessions per week. Since its only occurence and contribute negligible to the over metrix, I'll drop it. However I would log this as data quality issue for future monitoring.

In [8]:
session_bad_data = session_data[session_data.duplicated()]
session_data = session_data[~session_data.duplicated()]
session_data.duplicated().sum()

np.int64(0)

In [9]:
print("Total Rows with 'Unknown' Session Data:   ",session_data[session_data['session_start']=='unknown'].shape)

#Need to check these 'Unknown are random or pattern
unknown_sessions = session_data[session_data['session_start']=='unknown']
print("Total Unknown Session Users:  ",unknown_sessions['user_id'].nunique())
print("Unknown Session Users Value Counts: \n",unknown_sessions['user_id'].value_counts())
print("Result of Session Duration Districution:\n",unknown_sessions['duration_minutes'].describe())

Total Rows with 'Unknown' Session Data:    (65, 3)
Total Unknown Session Users:   61
Unknown Session Users Value Counts: 
 user_id
U1475    2
U1295    2
U1057    2
U1062    2
U1384    1
        ..
U1409    1
U1225    1
U1279    1
U1423    1
U1449    1
Name: count, Length: 61, dtype: int64
Result of Session Duration Districution:
 count     65.000000
mean      39.384615
std       33.089695
min        5.000000
25%       15.000000
50%       30.000000
75%       45.000000
max      120.000000
Name: duration_minutes, dtype: float64


##### Around 3.6% of session records had session_start = 'unknown'. Upon investigation, these records were distributed across 61 unique users with no significant clustering, and the session durations followed a normal distribution. This suggests the issue is random rather than systematic. Since session_start is critical for time-based analysis and cannot be reliably imputed, these records were dropped to maintain data quality. The issue was also noted as a data quality concern for future monitoring.

In [10]:
#Dropping sessions with 'unknown' session start
session_data = session_data[~(session_data["session_start"]=="unknown")]

#checking we still have unknown session start values
(session_data["session_start"]=="unknown").sum() # Gives 0 rows 

np.int64(0)

In [11]:
course_data['category'].unique() # No Inconsistent Data

array(['Competitive', 'Math', 'Coding', 'Tech'], dtype=object)

In [12]:
#Handling Inconsistent Data
print("User Data Describe:\n",users_data.describe())
print("\n\nEnrollment Data Values: ",enrollment_data["status"].unique())
print("\n\nSession Data Types:\n",session_data.dtypes)

#Since session start is not datetime so we need to convert
session_data["session_start"] = pd.to_datetime(session_data['session_start'],errors='coerce')
print("\n\nSession Data Types after conversion:\n",session_data.dtypes)
session_data = session_data[~session_data['session_start'].isnull()]
print("Remining Nulls: ",session_data["session_start"].isnull().sum())


User Data Describe:
               age  age_missing
count  500.000000   500.000000
mean    25.938000     0.058000
std      7.968258     0.233978
min     13.000000     0.000000
25%     19.000000     0.000000
50%     26.000000     0.000000
75%     33.000000     0.000000
max     39.000000     1.000000


Enrollment Data Values:  ['dropped' 'active' 'completed']


Session Data Types:
 user_id              object
session_start        object
duration_minutes    float64
dtype: object


Session Data Types after conversion:
 user_id                     object
session_start       datetime64[ns]
duration_minutes           float64
dtype: object
Remining Nulls:  0


## $\color{brown}{Retention Analysis}$ 

In [13]:
# Dealing with Users with 0 Session
enrolled_user = enrollment_data["user_id"].unique()

session_user = session_data["user_id"].unique()

user_zero_session = []

for user in enrolled_user:
    if user not in session_user:
        user_zero_session.append(user)

print(f"Out of {len(enrolled_user)} users, {len(user_zero_session)} have Zero Session  ")

Out of 410 users, 9 have Zero Session  


##### Most users at least try the platform — dropout happens after engagement starts

In [14]:
# Finding How many people are dropped among Zero Session Users
dropped_user= list(enrollment_data["user_id"][enrollment_data["status"]=="dropped"])
len_zero_session = len(user_zero_session)
len_zero_session_dropped = len(set(dropped_user) & set(user_zero_session))
print(f"Out of {len_zero_session}, {len_zero_session_dropped} are dropped out!")

# finding Dropout rate of Zero Session User 
zero_session_dropped = round(len_zero_session_dropped/len_zero_session *100, 2)
print("Dropped Out Ratio of User who never start Session: ", zero_session_dropped,"%")

Out of 9, 7 are dropped out!
Dropped Out Ratio of User who never start Session:  77.78 %


##### <font color="dark green">  Users with *Zero Sessions* show ~78% dropout rate, indicating a critical onboarding gap, though they represent a small portion of total users.

In [15]:
# Creating User level Data
user_level_data = enrollment_data[["user_id","status"]]

#grouping by user_id
user_level_data = user_level_data.groupby("user_id")["status"].apply(list).reset_index()

#function to decide final status of each user
def decide_status(x):
    if 'dropped' in x:
        return 'dropped'
    elif all(s=='completed' for s in x):
        return 'completed'
    else:
        return 'active'
        
user_level_data['final_status'] = user_level_data["status"].apply(decide_status)

# copying user level data with all status list for reference
user_level_data_copy = user_level_data.copy()

#getting only user with final status 
user_level_data = user_level_data[['user_id','final_status']]
print("Final User Level Data for Analysis:\n",user_level_data.head(10))

print("Dropped Users: ",(user_level_data["final_status"]=="dropped").sum())
print("Completed Users: ",(user_level_data["final_status"]=="completed").sum())
print("Active Users: ",(user_level_data["final_status"]=="active").sum())


Final User Level Data for Analysis:
   user_id final_status
0   U1001       active
1   U1002      dropped
2   U1003      dropped
3   U1004      dropped
4   U1006      dropped
5   U1007      dropped
6   U1008      dropped
7   U1010    completed
8   U1011      dropped
9   U1012    completed
Dropped Users:  235
Completed Users:  76
Active Users:  99


Dropout is very high
Roughly:
- Dropped ≈ ~57%
- Completed ≈ ~18%
- Active ≈ ~24%

“More than half of users are dropping out”
- This means:
Engagement problem ❗,
Course experience issue ❗,
Retention is weak ❗

### <font color="black">  Engagement Metrics 

In [16]:
# Calculating Engagement metrix
engagement_metrix_data = session_data.groupby("user_id").agg(session_count_per_user=("user_id","size"),total_time_spent=("duration_minutes","sum"),avg_duration=("duration_minutes","mean")).reset_index()


#getting final data by merging zero session users with users with sessions
user_level_data = user_level_data.merge(engagement_metrix_data,how="left")
print("\n\nUser level Data with Session Count:\n",user_level_data.head())

# Replacing 'NaN' values in zero session users to 0
user_level_data['session_count_per_user'] = user_level_data['session_count_per_user'].fillna(0)
user_level_data['total_time_spent'] = user_level_data['total_time_spent'].fillna(0)
user_level_data['avg_duration'] = user_level_data['avg_duration'].fillna(0)

# Checking Zero Session users in final user level data, user_id (U1153) has no session checked in previously
print("\n\nChecking Zero Session users in final user level data : \n",user_level_data[user_level_data["user_id"]=="U1153"])

#confirming we have 9 zero session users as we checked earlier
print("No. of Users with zero sessions: ",(user_level_data['session_count_per_user']==0).sum())



User level Data with Session Count:
   user_id final_status  session_count_per_user  total_time_spent  avg_duration
0   U1001       active                     3.0              45.0          15.0
1   U1002      dropped                     4.0             100.0          25.0
2   U1003      dropped                     2.0              75.0          37.5
3   U1004      dropped                     5.0              70.0          14.0
4   U1006      dropped                     2.0              90.0          45.0


Checking Zero Session users in final user level data : 
     user_id final_status  session_count_per_user  total_time_spent  \
127   U1153      dropped                     0.0               0.0   

     avg_duration  
127           0.0  
No. of Users with zero sessions:  9


In [17]:
user_level_data['avg_duration'] = round(user_level_data['avg_duration'],2)

print("The Final Dataset for Retention Analysis and Prediction: \n\n",user_level_data.head(10))

The Final Dataset for Retention Analysis and Prediction: 

   user_id final_status  session_count_per_user  total_time_spent  avg_duration
0   U1001       active                     3.0              45.0         15.00
1   U1002      dropped                     4.0             100.0         25.00
2   U1003      dropped                     2.0              75.0         37.50
3   U1004      dropped                     5.0              70.0         14.00
4   U1006      dropped                     2.0              90.0         45.00
5   U1007      dropped                     4.0             200.0         50.00
6   U1008      dropped                     1.0               5.0          5.00
7   U1010    completed                     4.0              95.0         23.75
8   U1011      dropped                     4.0             185.0         46.25
9   U1012    completed                     6.0             140.0         23.33


In [18]:
#Need to analyse data distribution in user level data to create buckets
print("User Level Data Distribution:\n",engagement_metrix_data.describe())

# Creating Buckets so that we analyse the dropput ratio
session_count_bucket = [[0,2,'0-2'],[3,4,'3-4'],[5,6,'5-6'],[7,float('inf'),'7+']]
total_time_spent_bucket = [[0,100,'0-100'],[101,200,'101-200'],[201,300,'201-300'],[301,float('inf'),'300+']]
avg_duration_bucket = [[0,25,'0-25'],[25,45,'25-45'],[45,65,'45-65'],[65,float('inf'),'65+']]

# Fuction to decide which bucket each row falls
def decide_bucket(value,bucket):
    for start,end,refer in bucket:
        if start<=value and value<=end:
            return refer
    return 'unknown'
    

# Adding user level data into each bucket
print("\n\nUser level Data after bucket allocation: ")
user_level_data["session_count_bucket"] = user_level_data["session_count_per_user"].apply(lambda x : decide_bucket(x,session_count_bucket))
user_level_data["total_time_spent_bucket"] = user_level_data["total_time_spent"].apply(lambda x: decide_bucket(x, total_time_spent_bucket))
user_level_data["avg_duration_bucket"] = user_level_data["avg_duration"].apply(lambda x: decide_bucket(x,avg_duration_bucket))
user_level_data.head()


User Level Data Distribution:
        session_count_per_user  total_time_spent  avg_duration
count               489.00000        489.000000    489.000000
mean                  3.95501        156.523517     39.474050
std                   1.98096        108.182545     21.568300
min                   1.00000          5.000000      5.000000
25%                   2.00000         70.000000     24.166667
50%                   4.00000        135.000000     35.714286
75%                   5.00000        220.000000     50.625000
max                  11.00000        570.000000    120.000000


User level Data after bucket allocation: 


,user_id,final_status,session_count_per_user,total_time_spent,avg_duration,session_count_bucket,total_time_spent_bucket,avg_duration_bucket
0,U1001,active,3.0,45.0,15.0,3-4,0-100,0-25
1,U1002,dropped,4.0,100.0,25.0,3-4,0-100,0-25
2,U1003,dropped,2.0,75.0,37.5,0-2,0-100,25-45
3,U1004,dropped,5.0,70.0,14.0,5-6,0-100,0-25
4,U1006,dropped,2.0,90.0,45.0,0-2,0-100,25-45


In [19]:
# Need to analyse value to status to find dropout ration : dropped = 1 , not_dropped=0

user_level_data['is_dropped'] = user_level_data["final_status"].apply(lambda x: 1 if x=='dropped' else 0)
print("User level data after defining dropout variable: \n",user_level_data.head(10))

User level data after defining dropout variable: 
   user_id final_status  session_count_per_user  total_time_spent  \
0   U1001       active                     3.0              45.0   
1   U1002      dropped                     4.0             100.0   
2   U1003      dropped                     2.0              75.0   
3   U1004      dropped                     5.0              70.0   
4   U1006      dropped                     2.0              90.0   
5   U1007      dropped                     4.0             200.0   
6   U1008      dropped                     1.0               5.0   
7   U1010    completed                     4.0              95.0   
8   U1011      dropped                     4.0             185.0   
9   U1012    completed                     6.0             140.0   

   avg_duration session_count_bucket total_time_spent_bucket  \
0         15.00                  3-4                   0-100   
1         25.00                  3-4                   0-100   
2       

### <font color="black">  Dropout Ratio

In [20]:
# For session_count
session_count_dropout = user_level_data.groupby('session_count_bucket').agg(total_users=("user_id","count"),dropped_user=("is_dropped","sum")).reset_index()

session_count_dropout['dropout_ratio (%)'] = round((session_count_dropout['dropped_user']/session_count_dropout['total_users']*100),2)
print("Engagement Metrics Session Count Dropout Ratio Data:")
session_count_dropout.head()

Engagement Metrics Session Count Dropout Ratio Data:


,session_count_bucket,total_users,dropped_user,dropout_ratio (%)
0,0-2,111,63,56.76
1,3-4,165,106,64.24
2,5-6,87,44,50.57
3,7+,47,22,46.81


##### <font color="dark green"> Users dropping out are not just low-engagement users. The highest dropout occurs at 3–4 sessions, indicating a critical retention gap after initial onboarding.

In [21]:
# For Total_time_spent

total_time_dropout = user_level_data.groupby("total_time_spent_bucket").agg(total_users = ("user_id","count"),dropped_users = ("is_dropped","sum")).reset_index()
total_time_dropout["dropout_ratio (%)"] = round((total_time_dropout["dropped_users"]/total_time_dropout["total_users"]*100),2)

print("Engagement Metrics total Time Spent Dropout Ratio Data:")
total_time_dropout.head()

Engagement Metrics total Time Spent Dropout Ratio Data:


,total_time_spent_bucket,total_users,dropped_users,dropout_ratio (%)
0,0-100,163,84,51.53
1,101-200,142,90,63.38
2,201-300,63,41,65.08
3,300+,42,20,47.62


##### <font color="dark green">Users with moderate engagement (3–6 sessions or 100–300 minutes) show the highest dropout, indicating a retention failure after initial adoption but before long-term commitment

In [22]:
# For Average duration

avg_duration_dropout = user_level_data.groupby("avg_duration_bucket").agg(total_users=("user_id","count"),dropped_users=("is_dropped","sum")).reset_index()
avg_duration_dropout["dropout_ratio (%)"] = round((avg_duration_dropout['dropped_users']/avg_duration_dropout['total_users']*100),2)

print("Engagement Metrics Average Duration Dropout Ratio Data:")
avg_duration_dropout.head(20)

Engagement Metrics Average Duration Dropout Ratio Data:


,avg_duration_bucket,total_users,dropped_users,dropout_ratio (%)
0,0-25,125,65,52.00
1,25-45,165,89,53.94
2,45-65,89,58,65.17
3,65+,31,23,74.19


##### <font color="dark green">While higher total engagement reduces dropout, users with longer average session durations exhibit higher dropout rates, indicating that binge-learning behavior negatively impacts retention.

### <font color="dark blue"> Final Core Insight:
##### <font color="purple"> Retention is driven by consistent engagement (more sessions and higher total time), while binge-like behavior (high average session duration) increases the likelihood of dropout.” 

### <font color="black">  Behavioral Segmentation

In [23]:
# Funtion to decide the feature segment
def decide_category(value,threshold):
    for start,end,category  in threshold:
        if start<=value and value <=end:
            return category
    else:
        return 'unknown'


In [24]:
# Defining Thresholds
session_count_threshold = [[0,2,'low'],[3,5,'medium'],[6,float('inf'),'high']]
total_time_threshold = [[0,100,'low'],[101,220,'medium'],[221,float('inf'),'high']]
avg_duration_threshold = [[0,25,'low'],[25,45,'medium'],[45,float('inf'),'high']]

user_feature_category = user_level_data[['user_id','session_count_per_user','total_time_spent','avg_duration']]

# Creating Feature Categories
user_feature_category["session_category"] = user_feature_category['session_count_per_user'].apply(lambda x: decide_category(x,session_count_threshold))
user_feature_category['time_category'] = user_feature_category['total_time_spent'].apply(lambda x: decide_category(x,total_time_threshold))
user_feature_category['duration_category'] = user_feature_category['avg_duration'].apply(lambda x: decide_category(x,avg_duration_threshold))

#user_feature_category['status'] = user_level_data['final_status']
user_feature_category = user_feature_category.merge(user_level_data[['user_id','final_status','is_dropped']],on='user_id')
                                                    
print("Feature Category Data : \n")
user_feature_category.head()



Feature Category Data : 



C:\Users\Avinash EM\AppData\Local\Temp\ipykernel_22364\2850943370.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  user_feature_category["session_category"] = user_feature_category['session_count_per_user'].apply(lambda x: decide_category(x,session_count_threshold))
C:\Users\Avinash EM\AppData\Local\Temp\ipykernel_22364\2850943370.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  user_feature_category['time_category'] = user_feature_category['total_time_spent'].apply(lambda x: decide_category(x,total_

,user_id,session_count_per_user,total_time_spent,avg_duration,session_category,time_category,duration_category,final_status,is_dropped
0,U1001,3.0,45.0,15.0,medium,low,low,active,0
1,U1002,4.0,100.0,25.0,medium,low,low,dropped,1
2,U1003,2.0,75.0,37.5,low,low,medium,dropped,1
3,U1004,5.0,70.0,14.0,medium,low,low,dropped,1
4,U1006,2.0,90.0,45.0,low,low,medium,dropped,1


In [25]:
# decinding segment
def decide_segment(user_data):  
    if user_data['session_category']=='high':
        return'Consistent User'
    elif user_data['duration_category']=='high' and user_data['session_category'] != 'high':
        return 'Binge User'
    elif user_data['session_category']=='low' and user_data['time_category']=='low':
        return 'Low Engagement'
    else:
        return 'Moderate Users'

# defining user behavioral segments
user_feature_category["user_segment"] = user_feature_category.apply(decide_segment,axis=1)
print("User Data with Behavioral Segments: \n")
user_feature_category.head()


User Data with Behavioral Segments: 



,user_id,session_count_per_user,total_time_spent,avg_duration,session_category,time_category,duration_category,final_status,is_dropped,user_segment
0,U1001,3.0,45.0,15.0,medium,low,low,active,0,Moderate Users
1,U1002,4.0,100.0,25.0,medium,low,low,dropped,1,Moderate Users
2,U1003,2.0,75.0,37.5,low,low,medium,dropped,1,Low Engagement
3,U1004,5.0,70.0,14.0,medium,low,low,dropped,1,Moderate Users
4,U1006,2.0,90.0,45.0,low,low,medium,dropped,1,Low Engagement


In [26]:
# finding dropout ratio of each behavioral segment 

behavioral_segment_dropout = user_feature_category.groupby('user_segment').agg(total_users=('user_segment','count'),dropped_users=('is_dropped','sum')).reset_index()
behavioral_segment_dropout['dropout_ratio (%)']=round((behavioral_segment_dropout['dropped_users']/behavioral_segment_dropout['total_users']*100),2)

print("Behavioural Segment Dropout Ratio: ")
behavioral_segment_dropout.head()

Behavioural Segment Dropout Ratio: 


,user_segment,total_users,dropped_users,dropout_ratio (%)
0,Binge User,89,62,69.66
1,Consistent User,88,47,53.41
2,Low Engagement,85,43,50.59
3,Moderate Users,148,83,56.08


### <font color="dark blue">📊 Behavioural Segment Dropout Analysis

##### <font color="purple">📌 Key Insights

Binge Users exhibit the highest dropout rate (70%), indicating that users with high-intensity but low-frequency engagement are most prone to churn. In contrast, Consistent and Low Engagement users show relatively lower dropout rates (50–53%), highlighting the importance of regular learning behavior over total consumption.

##### <font color="purple">📌 Observations
Binge Users (69.66%) have the highest dropout among all segments 🚨
Moderate Users (56.08%) show unstable behavior with moderately high dropout
Consistent Users (53.41%) perform better, indicating regular engagement improves retention
Low Engagement Users (50.59%) surprisingly have the lowest dropout among all segments

##### <font color="purple">📌 Business Interpretation

The results suggest that engagement pattern matters more than engagement volume.
Binge users tend to consume large amounts of content in fewer sessions, which can lead to cognitive overload and burnout, increasing the likelihood of dropout.

On the other hand, Low Engagement users progress slowly but steadily, avoiding overload, which may contribute to relatively better retention. Consistent users, who engage regularly, demonstrate more stable learning behavior and improved retention outcomes.

##### <font color="purple">📌 Recommendations
- Convert Binge Users → Consistent Users
      Break content into smaller, manageable chunks
      Introduce daily learning goals and streak-based systems
      Implement reminders to encourage frequent engagement
- Stabilize Moderate Users
      Use nudges and personalized recommendations to build consistency
- Strengthen Consistent Users
      Introduce gamification, rewards, and progress tracking to maintain engagement
- Support Low Engagement Users
      Provide beginner-friendly pathways and gentle nudges to increase activity without overwhelming them

### <font color="black">  Business Segmentation

#### Age Segmentation

In [27]:
# Defining age buckets
age_bucket = [[13,18,'Teen'] , [19,24,'Early Learners'],[25,30,'Professionals'],[31,float('inf'),'Mid Carrier']]

# assigning age category values to each user
users_data['age_bucket'] = users_data['age'].apply(lambda x: decide_bucket(x,age_bucket))

# creating data table to analyse age segmentation
age_segment_data = users_data[['user_id','age','age_bucket']]

# merging data with user categorical data 
user_age_segment_data = age_segment_data.merge(user_feature_category[['user_id','user_segment','is_dropped']],on='user_id')
print("User Data for Age Segmentation: ")
user_age_segment_data.head()

User Data for Age Segmentation: 


,user_id,age,age_bucket,user_segment,is_dropped
0,U1001,32.0,Mid Carrier,Moderate Users,0
1,U1002,27.0,Professionals,Moderate Users,1
2,U1003,23.0,Early Learners,Low Engagement,1
3,U1004,20.0,Early Learners,Moderate Users,1
4,U1006,19.0,Early Learners,Low Engagement,1


In [28]:
#  User Dropout analysis as per age segment
age_segment_dropout_data = user_age_segment_data.groupby('age_bucket').agg(total_user=('user_id','count'),dropped_user=('is_dropped','sum')).reset_index()
age_segment_dropout_data['dropout_ratio (%)'] = round((age_segment_dropout_data['dropped_user']/age_segment_dropout_data['total_user']*100),2)
print("Users Dropout ratio as per Age Segmentation : ")
age_segment_dropout_data.head()

Users Dropout ratio as per Age Segmentation : 


,age_bucket,total_user,dropped_user,dropout_ratio (%)
0,Early Learners,75,48,64.00
1,Mid Carrier,140,82,58.57
2,Professionals,101,55,54.46
3,Teen,94,50,53.19


In [29]:
# User Behavior analysis as per age segment
age_seg_behavior_data = user_age_segment_data.groupby(['age_bucket','user_segment']).agg(total_user=('user_id','count'),dropped_user=('is_dropped','sum')).reset_index()
age_seg_behavior_data['dropout_ratio (%)'] = round((age_seg_behavior_data['dropped_user']/age_seg_behavior_data['total_user']*100),2)
print("Users Dropout Analysis as per Age Segmentation and User Segmentation : ")
age_seg_behavior_data.head()

Users Dropout Analysis as per Age Segmentation and User Segmentation : 


,age_bucket,user_segment,total_user,dropped_user,dropout_ratio (%)
0,Early Learners,Binge User,15,10,66.67
1,Early Learners,Consistent User,15,10,66.67
2,Early Learners,Low Engagement,17,11,64.71
3,Early Learners,Moderate Users,28,17,60.71
4,Mid Carrier,Binge User,35,28,80.00


### <font color="dark blue"> Final Core Insight:
##### <font color="purple"> Early Learners (19–24) show the highest overall dropout (~64%), indicating that younger users are more prone to churn. Additionally, across all age groups, Binge behavior consistently leads to the highest dropout, especially in Mid Career users where it reaches ~80%.

#### <font color="black">  Course Level Dropout analysis

In [30]:
# We need Enrollment Data and Course Data
print("Enrollment Data: ")
print(enrollment_data.head())

print("Course Data: ")
print(course_data.head())

Enrollment Data: 
  user_id course_id enrollment_date     status
0   U1341      C203      2025-04-03    dropped
1   U1475      C205      2024-09-11    dropped
2   U1062      C206      2025-05-03     active
3   U1188      C209      2025-05-30  completed
4   U1359      C218      2024-12-24  completed
Course Data: 
  course_id                        course_name     category         level  \
0      C200  Special educational needs teacher  Competitive           NaN   
1      C201                          Mudlogger         Math      Advanced   
2      C202                       Sports coach       Coding  Intermediate   
4      C204             Sales professional, IT  Competitive  Intermediate   
5      C205              Pharmacist, community  Competitive      Beginner   

   rating  
0     1.0  
1     4.0  
2     NaN  
4     1.0  
5     1.0  


In [31]:
# getting required columns from course data
course_seg_data = course_data[['category','course_id']]

# merging course segment data to enrollment data
course_seg_data = course_seg_data.merge(enrollment_data[['course_id','user_id','status']])
course_seg_data = course_seg_data.merge(user_age_segment_data[['user_id','age_bucket','user_segment']])

# making status numerical for calculation
course_seg_data['is_dropped'] = course_seg_data['status'].apply(lambda x: 1 if x=='dropped' else 0)

# Need to get course categories with priority one
status_priority = { 
    'dropped':2,
    'completed':1,
    'active':0
}
course_seg_data['status_rank'] = course_seg_data['status'].map(status_priority)
print("The Course Level Dropout Data: (before) \n\n",course_seg_data.head(),course_seg_data.shape)
course_seg_data = course_seg_data.sort_values(['user_id','course_id','status_rank'],ascending=[False,False,False]).drop_duplicates(subset=['user_id','course_id'])
course_seg_data = course_seg_data[course_seg_data['status']!='active']
course_seg_data = course_seg_data.drop(columns="status_rank")
print("The Course Level Dropout Data: (shape)\n\n",course_seg_data.shape)
print('\n\nDuplicates:\n',course_seg_data[course_seg_data.duplicated(['user_id','course_id'], keep=False)])
course_seg_data.head()


The Course Level Dropout Data: (before) 

       category course_id user_id     status   age_bucket    user_segment  \
0  Competitive      C200   U1434     active  Mid Carrier  Moderate Users   
1  Competitive      C200   U1434  completed  Mid Carrier  Moderate Users   
2  Competitive      C200   U1379    dropped         Teen  Moderate Users   
3  Competitive      C200   U1177  completed  Mid Carrier      Binge User   
4  Competitive      C200   U1263  completed  Mid Carrier  Moderate Users   

   is_dropped  status_rank  
0           0            0  
1           0            1  
2           1            2  
3           0            1  
4           0            1   (670, 8)
The Course Level Dropout Data: (shape)

 (444, 7)


Duplicates:
 Empty DataFrame
Columns: [category, course_id, user_id, status, age_bucket, user_segment, is_dropped]
Index: []


,category,course_id,user_id,status,age_bucket,user_segment,is_dropped
498,Coding,C212,U1499,completed,Professionals,Binge User,0
285,Math,C208,U1499,dropped,Professionals,Binge User,1
76,Math,C201,U1498,completed,Teen,Consistent User,0
57,Math,C201,U1497,completed,Early Learners,Low Engagement,0
530,Tech,C215,U1496,dropped,Mid Carrier,Binge User,1


“All course enrollments where the user has reached a final outcome (either completed or dropped)”

In [32]:
# Dropout Ratio for each Course category
course_cat_dropout = course_seg_data.groupby('category').agg(dropped_enroll=('is_dropped','sum'),total_enroll=('is_dropped','count')).reset_index()
course_cat_dropout['dropout_ratio (%)'] = round((course_cat_dropout['dropped_enroll']/ course_cat_dropout['total_enroll']*100),2)
print("Course category Dropout Ratio: ")
course_cat_dropout.head()


Course category Dropout Ratio: 


,category,dropped_enroll,total_enroll,dropout_ratio (%)
0,Coding,42,96,43.75
1,Competitive,90,171,52.63
2,Math,63,120,52.50
3,Tech,34,57,59.65


##### <font color="purple">Competitive courses require immediate attention due to high enrollment and significant dropout (>50%), indicating a large-scale retention issue. However, further analysis is needed to determine whether this is driven by course design or user behavior.

In [33]:
# Course Category - User Segmentation Dropout Ratio
course_cat_seg_dropout = course_seg_data.groupby(['category','user_segment']).agg(dropped_enroll=('is_dropped','sum'),total_enroll=('is_dropped','count')).reset_index()
course_cat_seg_dropout['dropout_ratio (%)'] = round((course_cat_seg_dropout['dropped_enroll']/ course_cat_seg_dropout['total_enroll']*100),2)
print("Course category - User Segmentatio Dropout Ratio: ")

course_cat_seg_dropout.head(20)



Course category - User Segmentatio Dropout Ratio: 


,category,user_segment,dropped_enroll,total_enroll,dropout_ratio (%)
0,Coding,Binge User,11,23,47.83
1,Coding,Consistent User,11,22,50.00
2,Coding,Low Engagement,5,17,29.41
3,Coding,Moderate Users,15,34,44.12
4,Competitive,Binge User,21,38,55.26
5,Competitive,Consistent User,13,25,52.00
6,Competitive,Low Engagement,16,35,45.71
7,Competitive,Moderate Users,40,73,54.79
8,Math,Binge User,17,25,68.00
9,Math,Consistent User,9,20,45.00


### <font color="dark blue"> Final Core Insight:
##### <font color="purple">Competitive courses exhibit consistently high dropout across all user segments, including engaged users, indicating a systemic issue in course design or difficulty, while categories like Math show behavior-driven dropout primarily among binge learners.
### <font color="red"> Attention! 
##### <font color="purple">Age-based segmentation was not prioritized as it introduced data sparsity without significantly improving actionable insights beyond behavior and category-level patterns.

### <font color="dark blue"> Dataset for Dashboard:

In [34]:
# getting Engagement metrics table
print("Getting Data of Engagement Metrics: \n",user_level_data.head())
print("Getting Data for Course and User details: \n",course_seg_data.head())

# Merging two to get desired dataset for dashboard creation
final_dashboard_data = course_seg_data.merge(user_level_data[['user_id','session_count_per_user','total_time_spent','avg_duration']],on='user_id')
print("\nChecking for Duplicated: ",final_dashboard_data.duplicated().sum())
final_dashboard_data.head()


Getting Data of Engagement Metrics: 
   user_id final_status  session_count_per_user  total_time_spent  \
0   U1001       active                     3.0              45.0   
1   U1002      dropped                     4.0             100.0   
2   U1003      dropped                     2.0              75.0   
3   U1004      dropped                     5.0              70.0   
4   U1006      dropped                     2.0              90.0   

   avg_duration session_count_bucket total_time_spent_bucket  \
0          15.0                  3-4                   0-100   
1          25.0                  3-4                   0-100   
2          37.5                  0-2                   0-100   
3          14.0                  5-6                   0-100   
4          45.0                  0-2                   0-100   

  avg_duration_bucket  is_dropped  
0                0-25           0  
1                0-25           1  
2               25-45           1  
3                0-25   

,category,course_id,user_id,status,age_bucket,user_segment,is_dropped,session_count_per_user,total_time_spent,avg_duration
0,Coding,C212,U1499,completed,Professionals,Binge User,0,4.0,225.0,56.25
1,Math,C208,U1499,dropped,Professionals,Binge User,1,4.0,225.0,56.25
2,Math,C201,U1498,completed,Teen,Consistent User,0,7.0,570.0,81.43
3,Math,C201,U1497,completed,Early Learners,Low Engagement,0,2.0,60.0,30.00
4,Tech,C215,U1496,dropped,Mid Carrier,Binge User,1,4.0,255.0,63.75


In [35]:
# Reading this final_dashboard_data to csv file for visualization
final_dashboard_data.to_csv("Output/edumax_final_analysis_dashboard_data.csv",index=False)